receiver/prometheusexecreceiver/receiver_test.go

```diff
@@ -0,0 +1,126 @@
+	execErrorTest(t, config.Receivers["prometheus_exec"])
+
+	// Normal test with port undefined by user
+	endToEndScrapeTest(t, config.Receivers["prometheus_exec/end_to_end_test/2"], "end-to-end port not defined")
+}
+
+// execErrorTest makes sure the config passed throws an error, since it's missing the exec key
+func execErrorTest(t *testing.T, errorReceiverConfig configmodels.Receiver) {
+	_, err := new(component.ReceiverCreateParams{Logger: zap.NewNop()}, errorReceiverConfig.(*Config), nil)
+	if err == nil {
+		t.Errorf("end_to_end_test.go didn't get error, was expecting one since this config has no 'exec' key")
+	}
+}
+
+// endToEndScrapeTest scrapes a test endpoint (test_prometheus_exporter.go) twice, and between each scrape yields the execution with Sleep() to wait for the subprocess (exporter) to restar
+// - wait time is about 1s - to fail and restart, meaning it verifies three things: the scrape is successful (twice), the process was restarted correctly when failed and the underlying
+// Prometheus receiver was correctly stopped and then restarted. For extra testing the metrics values are different every time the subprocess exporter is started
+// And the uniqueness of the metric scraped is verified
+func endToEndScrapeTest(t *testing.T, receiverConfig configmodels.Receiver, testName string) {
+	sink := &exportertest.SinkMetricsExporter{}
+	wrapper, err := new(component.ReceiverCreateParams{Logger: zap.NewNop()}, receiverConfig.(*Config), sink)
+	if err != nil {
+		t.Errorf("end_to_end_test.go got error = %w", err)
+	}
+
+	ctx := context.Background()
+	err = wrapper.Start(ctx, componenttest.NewNopHost())
+	if err != nil {
+		t.Errorf("end_to_end_test.go got error = %w", err)
+	}
+	defer wrapper.Shutdown(ctx)
+
+	var metrics []pdata.Metrics
+
+	// Make sure a first scrape works by checking for metrics in the test metrics exporter "sink", only return true when there are metrics
+	const waitFor = 15 * time.Second
+	const tick = 500 * time.Millisecond
+	require.Eventuallyf(t, func() bool {
+		got := sink.AllMetrics()
+		return len(got) != 0
+	}, waitFor, tick, "No metrics were collected after %v for the first scrape (%v)", waitFor, testName)
+
+	metrics = sink.AllMetrics()
+
+	// Make sure the second scrape is successful, and validate that the metrics are different in the second scrape
+	require.Eventuallyf(t, func() bool {
+		got := sink.AllMetrics()
+		if len(got) == 0 || len(got) == len(metrics) {
+			return false
+		}
+		if validateMetrics(&got) {
+			return true
+		}
+
+		metrics = got
+		return false
+	}, waitFor, tick, "No metrics were collected after %v for the second scrape (%v)", waitFor, testName)
+}
+
+// validateMetrics iterates over the found metrics and returns true if it finds at least 2 unique metrics, meaning the endpoint
+// was successfully scraped twice AND the subprocess being handled was stopped and restarted
+func validateMetrics(metricsSlice *[]pdata.Metrics) bool {
```

Comment:
Why use a pointer to a slice here?

Reply:
I don't really need a copy of the data, I'm just analyzing it in that function, so this would increase performance, no? Rather than copying over the entire slice

Reply:
A slice already contains a pointer to the underlying array - see https://blog.golang.org/slices-intro#:~:text=A%20slice%20is%20a%20descriptor,referred%20to%20by%20the%20slice.